### ISSUE TITLE
Uncaught TypeError: context2 is undefined

### ISSUE THREAD

**Issue by crimson-med** [straker/kontra#319]
I'm currently working on a game using kontra  and i'm using typescript however when I bundle everything together i'm getting an issue with kontra.

```
Uncaught TypeError: context2 is undefined
    _p kontra.mjs:769
    init kontra.mjs:727
    Updatable kontra.mjs:384
    GameObject kontra.mjs:440
    Text kontra.mjs:711
    factory$7 kontra.mjs:815
    <anonymous> text.js:2
[kontra.mjs:769:4](http://localhost:8081/kontra/kontra.mjs)
    _p kontra.mjs:769
    init kontra.mjs:727
    Updatable kontra.mjs:384
    GameObject kontra.mjs:440
    Text kontra.mjs:711
    factory$7 kontra.mjs:815
    <anonymous> text.js:2
```
    
The line pointed out is the following:

```
context2.font = this.font;
```

Situated in the `_p()` of the `class Text extends GameObject {`

This error happens for the following Text entities as well as a Sprite.

test.ts

```ts
import { Text } from "kontra"
//Set our title
export let title = Text({
    text: 'GIVE ME SPACE!',
    font: '20px Luminari, fantasy',
    color: 'white',
    x: 100,
    y: 200,
    anchor: { x: 0.5, y: 0.5 },
    textAlign: 'center'
});
export let score = Text({
    text: 'Score:',
    font: '20px Luminari, fantasy',
    color: 'white',
    x: 50 + 16 * 3,
    y: 150,
    anchor: { x: 0.5, y: 0.5 },
    textAlign: 'center'
});
```

The issue comes when rendering.

index.ts

```ts
let loop = GameLoop({  // create the main game loop
  render: () => {
    // Non need to render the collision layer
    if (this.tileEngine) {
        this.tileEngine.renderLayer("ground");
    }
    // character.render();
    title.render();
    score.render();
    this.lifeCollection.forEach(element => {
        //console.log(element)
        element.render();
    });
  }
});

```

The issue seems to be when sprites are defined in other files. the render then can't pass the context properly.

Is there a way to define sprites in another file and then handle the render properly ?

**Comment by straker** [comment]
So this might be related to #318. If `text.ts` get's imported before `index.ts` is able to call `init()` to pass the canvas and setup the `context`, then when `Text` is created it doesn't have the context set so it won't be defined when it renders.

**Comment by crimson-med** [comment]
@straker in my current setup what I tried to do was the following

- src
---sprites
------character.ts
------title.ts
------score.ts
---index.ts

Where index.ts was a class and in the constructor I initialized the tileengine and the tileloop.

I came through two problems with that structure. which I think are a bit similar

the sprites can't work properly as they are imported in index.ts at the top and since the `Sprite()` is already called in the file itself it has already initialized without the context.

I tries passing the context to the sprite thinking that could be a good solution but it didn't work.

The second one is the fact that the context is auto passed to the `update` and `render` function. 
I wanted to put those functions in a seperate method in the class for cleaner code and then having

```
update(this.updateState)
```

this way I could easily access properties defined in my class butt this means in loose the binding to the context which is used in render. I tried passing the context to the `updateState` function so I could try something like `character.render(context)` but this was also with no avail.

for #318 do you have any  idea yet?

**Comment by straker** [comment]
Not at the moment. I need to go through everything to see what options I have.

**Comment by crimson-med** [comment]
> Not at the moment. I need to go through everything to see what options I have.

Alright I understand! Looking forward if you do find a solution. I love the simplicity of kontra and how lightweight it is. Currently using it for a mini mmo type tileset game and hopefully ill be able to share it soon

**Comment by straker** [comment]
Alright, I think I fixed this in #332. Should be able to create any object now before calling `init` and use them so long as you don't try to render them before calling `init`.

**Comment by crimson-med** [comment]
Awesome will give it a try in the coming weeks!!! Thanks a lot!
